src/include/btree_inline.h

```diff
@@ -2024,50 +2024,38 @@ __wt_bt_col_var_cursor_walk_txn_read(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *
  *     Return if the cursor is pointing to a page with deleted records and can be skipped for cursor
  *     traversal.
  */
-static inline bool
-__wt_btcur_skip_page(WT_CURSOR_BTREE *cbt)
+static inline int
+__wt_btcur_skip_page(WT_SESSION_IMPL *session, WT_REF *ref, void *context, bool *skipp)
 {
     WT_ADDR_COPY addr;
-    WT_PAGE *page;
-    WT_REF *ref;
-    WT_SESSION_IMPL *session;
     uint8_t previous_state;
-    bool can_skip;
 
-    session = CUR2S(cbt);
-    ref = cbt->ref;
-    page = cbt->ref == NULL ? NULL : cbt->ref->page;
+    *skipp = false; /* Default to reading */
 
-    if (page == NULL)
-        return false;
+    WT_UNUSED(context);
 
-    previous_state = ref->state;
-    can_skip = false;
+    if (ref == NULL)
+        return (0);
 
     /*
      * Determine if all records on the page have been deleted and all the tombstones are visible to
      * our transaction. If so, we can avoid reading the records on the page and move to the next
      * page. We base this decision on the aggregate stop point added to the page during the last
-     * reconciliation. We can skip this test if the page has been modified since it was reconciled
-     * or the underlying cursor is configured to ignore tombstones.
+     * reconciliation. We can skip this test if the page has been modified since it was reconciled.
      *
      * We are making these decisions while holding a lock for the page as checkpoint or eviction can
-     * make changes to the data structures (i.e., aggregate timestamps) we are reading.
+     * make changes to the data structures (i.e., aggregate timestamps) we are reading. It is okay
+     * if the page is not in memory, or gets evicted before we lock it. In such a case, we can forgo
+     * checking if the page has been modified. So, only do a page modified check if the page was in
+     * memory before locking.
      */
-    if (session->txn->isolation == WT_ISO_SNAPSHOT && !__wt_page_is_modified(page) &&
-      !F_ISSET(&cbt->iface, WT_CURSTD_IGNORE_TOMBSTONE) && previous_state == WT_REF_MEM) {
-
-        /* We only try to lock the page once. */
-        if (!WT_REF_CAS_STATE(session, ref, previous_state, WT_REF_LOCKED))
-            return false;
+    WT_REF_LOCK(session, ref, &previous_state);
+    if ((previous_state != WT_REF_MEM || ref->page == NULL || !__wt_page_is_modified(ref->page)) &&
```

Comment:
@sulabhM, shouldn't this be `ref->page != NULL`?

Reply:
No, It is saying that if the page is NULL, don't have to proceed with the modified check, and instead can straight look at the aggregated address. So the page is modified check will only be performed if "ref->page == NULL" is false, ie "ref->page != NULL".

Reply:
Ah yes. Sorry I didn't realize that there is a "OR" condition there.

Reply:
@sulabhM, @hbokhari, I think there is a legitimate question being raised here. There are 5 possible `WT_REF.state` values: DISK, DELETED, LOCKED, MEM and SPLIT. So, what are the possible `previous_state` values?

* WT_REF_DISK    yes, do the test
* WT_REF_DELETED     yes, do the test
* WT_REF_LOCKED    can't happen, that would be a bad bug somewhere
* WT_REF_MEM    yes, do the test ONLY if the page hasn't been modified
* WT_REF_SPLIT    I am pretty confident this can't happen because the traversal code will have already dealt with split pages, but it's not as clear as other paths.

I think it might be a little safer (and future proof the code in case a new `WT_REF.state` was added in the future), and possibly add some clarity, to write the test this way:

```
if ((previous_state == WT_REF_DISK || previous_state == WT_REF_DELETED ||
    (previous_state == WT_REF_MEM && !__wt_page_is_modified(ref->page))) ...
```

I'm not testing `ref.page == NULL`, because if the page is in memory, the `ref.page` pointer has to be valid.

@sulabhM, this is an untested change, and I think the current code is actually OK, I just think this would add a bit of safety and maybe clarity to the code. Your call, and if you are good with the current code, that's fine with me.